apps/emqx/test/emqx_schema_tests.erl

```diff
@@ -178,27 +178,290 @@ ssl_opts_gc_after_handshake_test_not_rancher_listener_test() ->
+                [{"host1", DefaultPort}, {"host2", DefaultPort}],
+                Parse(["host1", "host2"])
+            )
+        ),
+        ?T(
+            "multiple servers, list(binary), no port",
+            ?assertEqual(
+                [{"host1", DefaultPort}, {"host2", DefaultPort}],
+                Parse([<<"host1">>, <<"host2">>])
+            )
+        ),
+        ?T(
+            "multiple servers, string, with port",
+            ?assertEqual(
+                [{"host1", 1234}, {"host2", 2345}],
+                Parse("host1:1234, host2:2345")
+            )
+        ),
+        ?T(
+            "multiple servers, binary, with port",
+            ?assertEqual(
+                [{"host1", 1234}, {"host2", 2345}],
+                Parse(<<"host1:1234, host2:2345, ">>)
+            )
+        ),
+        ?T(
+            "multiple servers, list(string), with port",
+            ?assertEqual(
+                [{"host1", 1234}, {"host2", 2345}],
+                Parse([" host1:1234 ", "host2:2345"])
+            )
+        ),
+        ?T(
+            "multiple servers, list(binary), with port",
+            ?assertEqual(
+                [{"host1", 1234}, {"host2", 2345}],
+                Parse([<<"host1:1234">>, <<"host2:2345">>])
+            )
+        ),
+        ?T(
+            "unexpected multiple servers",
+            ?assertThrow(
+                "expecting_one_host_but_got: 2",
+                emqx_schema:parse_server(<<"host1:1234, host2:1234">>, #{default_port => 1})
+            )
+        ),
+        ?T(
+            "multiple servers without ports invalid string list",
+            ?assertThrow(
+                "hostname_has_space",
+                Parse2(["host1 host2"], #{no_port => true})
+            )
+        ),
+        ?T(
+            "multiple servers without ports invalid binary list",
+            ?assertThrow(
+                "hostname_has_space",
+                Parse2([<<"host1 host2">>], #{no_port => true})
+            )
+        ),
+        ?T(
+            "multiple servers wihtout port, mixed list(binary|string)",
+            ?assertEqual(
+                ["host1", "host2"],
+                Parse2([<<"host1">>, "host2"], #{no_port => true})
+            )
+        ),
+        ?T(
+            "no default port, missing port number in config",
+            ?assertThrow(
+                "missing_port_number",
+                emqx_schema:parse_server(<<"a">>, #{})
+            )
+        ),
+        ?T(
+            "empty binary string",
+            ?assertEqual(
+                undefined,
+                emqx_schema:parse_server(<<>>, #{no_port => true})
+            )
+        ),
+        ?T(
+            "empty array",
+            ?assertEqual(
+                undefined,
+                emqx_schema:parse_servers([], #{no_port => true})
+            )
+        ),
+        ?T(
+            "empty binary array",
+            ?assertThrow(
+                "bad_host_port",
```

Comment:
I thought this would lead to undefined.  :thinking:

Reply:
arrar elements are not filtered.

Reply:
translating empty string or binary to undefined is mostly to support un-setting from env override.

Reply:
because there is no way to *delete* using overrides.